src/mon/MgrStatMonitor.cc

```diff
@@ -0,0 +1,200 @@
+
+  const pool_stat_t* get_pool_stat(int poolid) const override {
+    auto i = digest.pg_pool_sum.find(poolid);
+    if (i != digest.pg_pool_sum.end()) {
+      return &i->second;
+    }
+    return nullptr;
+  }
+
+  const pool_stat_t& get_pg_sum() const override { return digest.pg_sum; }
+  const osd_stat_t& get_osd_sum() const override { return digest.osd_sum; }
+
+  const osd_stat_t *get_osd_stat(int osd) const override {
+    auto i = digest.osd_stat.find(osd);
+    if (i == digest.osd_stat.end()) {
+      return nullptr;
+    }
+    return &i->second;
+  }
+  const mempool::pgmap::unordered_map<int32_t,osd_stat_t> &get_osd_stat() const override {
+    return digest.osd_stat;
+  }
+
+  size_t get_num_pg_by_osd(int osd) const override {
+    return digest.get_num_pg_by_osd(osd);
+  }
+
+  ceph_statfs get_statfs() const override {
+    return digest.get_statfs();
+  }
+
+  void print_summary(Formatter *f, ostream *out) const override {
+    digest.print_summary(f, out);
+  }
+  void dump_info(Formatter *f) const override {
+    digest.dump(f);
+  }
+  void dump_fs_stats(stringstream *ss,
+		     Formatter *f,
+		     bool verbose) const override {
+    digest.dump_fs_stats(ss, f, verbose);
+  }
+  void dump_pool_stats(const OSDMap& osdm, stringstream *ss, Formatter *f,
+		       bool verbose) const override {
+    digest.dump_pool_stats_full(osdm, ss, f, verbose);
+  }
+};
+
+
+#define dout_subsys ceph_subsys_mon
+#undef dout_prefix
+#define dout_prefix _prefix(_dout, mon)
+static ostream& _prefix(std::ostream *_dout, Monitor *mon) {
+  return *_dout << "mon." << mon->name << "@" << mon->rank
+		<< "(" << mon->get_state_name()
+		<< ").mgrstat ";
+}
+
+MgrStatMonitor::MgrStatMonitor(Monitor *mn, Paxos *p, const string& service_name)
+  : PaxosService(mn, p, service_name),
+    pgservice(new MgrPGStatService(digest))
+{
+}
+
+MgrStatMonitor::~MgrStatMonitor() = default;
+
+PGStatService *MgrStatMonitor::get_pg_stat_service()
+{
+  return pgservice.get();
+}
+
+void MgrStatMonitor::create_initial()
+{
+  dout(10) << dendl;
+  version = 0;
+}
+
+void MgrStatMonitor::update_from_paxos(bool *need_bootstrap)
+{
+  version = get_last_committed();
+  dout(10) << " " << version << dendl;
+  bufferlist bl;
+  get_version(version, bl);
+  if (version) {
+    assert(bl.length());
+    auto p = bl.begin();
+    bufferlist digestbl;
+    ::decode(digestbl, p);
+    auto q = digestbl.begin();
+    ::decode(digest, q);
+    ::decode(health_summary, p);
+    ::decode(health_detail, p);
+  }
+}
+
+void MgrStatMonitor::create_pending()
+{
+  dout(10) << " " << version << dendl;
+  pending_digest = digest;
+  pending_health_summary = health_summary;
+  pending_health_detail = health_detail;
+}
+
+void MgrStatMonitor::encode_pending(MonitorDBStore::TransactionRef t)
+{
+  ++version;
+  dout(10) << " " << version << dendl;
+  bufferlist digestbl, bl;
+  ::encode(pending_digest, digestbl, mon->get_quorum_con_features());
+  ::encode(digestbl, bl);
```

Comment:
@liewegas, the commit adding this digestbl says it's "just for bigbang's benefit", although I'm not sure how. What's up, and should it actually go int?